### PR TITLE
peer and metadata: Implement the Stringer interface for Peer and Metadata

### PR DESCRIPTION
**Pull request by AnomalRoil**
Effectively allowing to print contextes containing a `Peer` properly.
Instead of printing with `%v` as:
```
context.Background.WithValue(type peer.peerKey, val <not Stringer>)
```
contexts containing a peer will now print as:
```
context.Background.WithValue(type peer.peerKey, val Peer{Addr: '1.2.3.4:1234', AuthInfo: 'testAuthInfo'})
```

This is also satisfying Stringer interface for `Metadata`, allowing to print them easily too. Since `Metadata` are basically a map, notice that `String` output is not stable. I could change it to print it in alphabetical order if you prefer.

RELEASE NOTES:
- peer/peer: implement the Stringer interface for pretty printing Peer
- metadata/metadata: implement the Stringer interface for pretty printing MD

### CODE REVIEW

File: metadata/metadata.go
Comment (aranjans) [inline metadata/metadata.go:99]: Probably we could avoid this variable and just check for the base value("MD{")

File: peer/peer_test.go
Comment (aranjans) [inline peer/peer_test.go:39]: How about implementing end to end testing here?

**Review by aranjans — CHANGES_REQUESTED** [review]

File: peer/peer_test.go
Comment (AnomalRoil) [reply peer/peer_test.go:39]: I'm not sure I understand what you mean, do you have any example test I could look at doing so?

File: peer/peer.go
Comment (AnomalRoil) [reply peer/peer.go:57]: The `nil` checks are there to avoid risks of `nil` deref when printing/logging, which aren't avoided with the proposed alternative.

A more readable alternative would be to print all fields but put `<nil>` when these are `nil`, what do you prefer?

File: metadata/metadata.go
Comment (AnomalRoil) [reply metadata/metadata.go]: Done using the length of the builder.

File: balancer_wrapper.go
Comment (AnomalRoil) [inline balancer_wrapper.go:266]: The impossibility of finding out which `Addr` a given `SubConn` my balancer is using is something I stumbled upon today, so not sure if this PR is the right place for it, but since it's also related to a Stringer, I figured maybe?
 
But this requires a mutex lock I believe, because  `UpdateAddresses` can `ac.resetTransport` and that can change `curAddr.`... So not sure if it's worth it, especially in a `String`? 

I'd love to know how one can find out the exact `Addr` a given `balancer.SubConn` is using if you know how to do it, since I see  @townba removed `curAddr` from the `channelz.Infof` in `ac.updateAddrs` earlier this year.

Comment (AnomalRoil) [reply balancer_wrapper.go]: I have now discovered I could run a localhost Server and query the channelz directly using `grpc_channelz_v1`, which means I can query the target of a subchannel using this, I've reverted my commit since I don't feel good about adding a mutex lock in a `Stringer`.

File: metadata/metadata.go
Comment (arvindbr8) [inline metadata/metadata.go:100]: Not sure why do we need this conditional, removing this still makes the tests pass

File: metadata/metadata.go
Comment (arvindbr8) [inline metadata/metadata.go:96]: Curious, Is there a reason for using `{` here? The go standard for printing maps seems to have `[` instead.

File: metadata/metadata_test.go
Comment (arvindbr8) [inline metadata/metadata_test.go:345]: why not an exact string match?

File: peer/peer.go
Comment (arvindbr8) [inline peer/peer.go:49]: Is this codepath possible?

File: peer/peer.go
Comment (arvindbr8) [inline peer/peer.go:52]: mm, wouldnt the peer always have the peer's Addr?

File: peer/peer.go
Comment (arvindbr8) [inline peer/peer.go:57]: Why exclude `LocalAddr`?

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:71]: can you move each testCase into a `t.Run()`?

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:132]: I'm not sure if i understand this pattern here. Why dont we just create a Peer object and test p.String to tc.want

and testCases should just be 

```go
testCases := []struct {
	p	peer 
	want	string
}{
```

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:133]: This can also another test case above

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:139]: This can be a separate test case

File: metadata/metadata.go
Comment (arvindbr8) [inline metadata/metadata.go:93]: Please line wrap this to 80 cols

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
The change overall looks good to me. I took a pass at it. please take a look

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
The change overall looks good to me. I took a pass at it. please take a look

File: peer/peer.go
Comment (AnomalRoil) [reply peer/peer.go:49]: The goal here is to satisfy the `Stringer` interface, so it's true that someone doing 
```
var p *grpc.Peer
fmt.Println(p)
```
would correctly get `nil` as an output.

But a developer doing:
```
var p *grpc.Peer
p.String()
```
shouldn't be causing a `nil` dereference, or would you prefer if it would?
Here's a mocked example: https://go.dev/play/p/05c8CQF2EjM

I can replace the default `Peer<nil>` with just `"nil"` that being said, to match the usual Go pattern if you prefer?

File: peer/peer.go
Comment (AnomalRoil) [reply peer/peer.go:52]: The documentation didn't really said and I wanted to make sure to avoid pathological cases such as the above one, where someone creates an empty `Peer` and calls `String` on it.

File: peer/peer.go
Comment (AnomalRoil) [reply peer/peer.go:62]: I missed it entirely! Patched now.

File: metadata/metadata_test.go
Comment (AnomalRoil) [reply metadata/metadata_test.go:345]: Because of the fact that ranging over maps is non-deterministic, therefore the string representation of a map is not stable.

As mentioned above, happy to  change it to print it in alphabetical order of the keys if you prefer.

File: metadata/metadata.go
Comment (AnomalRoil) [reply metadata/metadata.go:100]: This is to avoid having a last trailing `","`, so we have `"MD{k1=[v1]}"` instead of `"MD{k1=[v1,]}"`.

Comment (arvindbr8) [reply metadata/metadata.go:101]: I'm still confused. Isn't that logic already done in `strings.Join(v, ", ")` below?

File: metadata/metadata.go
Comment (AnomalRoil) [reply metadata/metadata.go:96]: The curly braces `{}` felt more natural for a map, and the regular braces `[]` felt more like an array to me. 
I guess because it feels more like json then, or more like structs?  But I used `=` instead of `:` too.

This is the current formatting in the PR:
```
"MD{k1=[v1, v2], k2=[], k3=[1, 2, 3]}"
``` 

It's true that when doing a `fmt.Println(m)` for `m` a map, it will print as `"map[k:v k:v]"`, would you prefer that format? 
```
"MD[k1:[v1, v2] k2:[] k3:[1, 2, 3]]"
```

File: peer/peer.go
Comment (arvindbr8) [reply peer/peer.go:52]: I agree, the stringer should handle all the "pathological cases". For eg, what if this is called from a unit test without Addr set.

File: peer/peer.go
Comment (arvindbr8) [reply peer/peer.go:49]: I'm okay with this diff. Thanks!

File: metadata/metadata.go
Comment (arvindbr8) [reply metadata/metadata.go:97]: On second thoughts, i think this format is just fine. unless other maintainers have a strong opinion here. Thanks for clarifying.

File: metadata/metadata_test.go
Comment (arvindbr8) [reply metadata/metadata_test.go:345]: This is fine. I didnt mean to have a exact string matcher test here. Nobody should be relying on the output of a stringer method. 

But I think the only test we want is to make sure that md.String() does not cause a `panic`.  No further action required in this comment. Thanks!

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:39]: This unit test doesnt seem to be related to this particular patch-set. It's nice to have more coverage, but I would prefer this as part of a different PR, please.

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:126]: nit: 

add a field `name` to testCases, and use that here instead?

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:142]: please: 

```suggestion
	want := "context.Background.WithValue(type peer.peerKey, val Peer{Addr: '1.2.3.4:1234', LocalAddr: <nil>, AuthInfo: 'testAuthInfo-3'})"
	if got := fmt.Sprintf("%v", ctx); got !=  want {
		t.Fatalf("Unexpected stringer output, Got: %q; Want: %q", got, want)
```

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:134]:
```suggestion
			if got := p.String(); got != tc.want {
				t.Fatalf("Unexpected stringer output, got %q; want: %q", got, tc.want)
			}
```

per golang styleguide, got before want always. see: https://google.github.io/styleguide/go/decisions#got-before-want

File: metadata/metadata_test.go
Comment (arvindbr8) [inline metadata/metadata_test.go:356]:
```suggestion
				t.Fatalf("Metadata string %q is missing %q", strMd, want)
```

File: metadata/metadata_test.go
Comment (arvindbr8) [inline metadata/metadata_test.go:353]:
```suggestion
		got := test.md.String()
```

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Thanks, I've responded to our conversations here. Please take a look

File: metadata/metadata_test.go
Comment (AnomalRoil) [inline metadata/metadata_test.go:356]:
```suggestion
			if !strings.Contains(got, want) {
				t.Fatalf("Metadata string '%s' is missing '%s'", got, want)
```

File: peer/peer_test.go
Comment (AnomalRoil) [reply peer/peer_test.go:86]: TIL 👍🏻

File: peer/peer_test.go
Comment (arvindbr8) [inline peer/peer_test.go:53]: Just FYI, go `test` std library already does something similar by replacing `\s`s with `_`s.

File: metadata/metadata.go
Comment (arvindbr8) [reply metadata/metadata.go:101]: thanks for clarifying.

**Review by arvindbr8 — APPROVED** [review]

**Comment by arvindbr8** [comment]
Per policy we need one more review before we can check this in. I will ping a co-maintainer to take a look at this.

File: peer/peer_test.go
Comment (AnomalRoil) [reply peer/peer_test.go:53]: I was trying to convey information about whether the test case included `+` or didn't include `-` that specific field.

**Comment by serathius** [comment]
Making `String()` non-deterministic broke etcd stream selection mechanism: https://github.com/etcd-io/etcd/blob/21e5876f7f0539509c277b4c2a3ba1b1599b1721/client/v3/watch.go#L1038

**Comment by dims** [comment]
@serathius @zasweq the following patch to make the string deterministic solves our issue:

```
diff --git a/vendor/google.golang.org/grpc/metadata/metadata.go b/vendor/google.golang.org/grpc/metadata/metadata.go
index 6c01a9b359cc..ec1055f6ccf0 100644
--- a/vendor/google.golang.org/grpc/metadata/metadata.go
+++ b/vendor/google.golang.org/grpc/metadata/metadata.go
@@ -24,6 +24,7 @@ package metadata // import "google.golang.org/grpc/metadata"
 import (
 	"context"
 	"fmt"
+	"sort"
 	"strings"
 
 	"google.golang.org/grpc/internal"
@@ -95,11 +96,16 @@ func Pairs(kv ...string) MD {
 func (md MD) String() string {
 	var sb strings.Builder
 	fmt.Fprintf(&sb, "MD{")
-	for k, v := range md {
+	keys := make([]string, 0, len(md))
+	for k := range md {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
 		if sb.Len() > 3 {
 			fmt.Fprintf(&sb, ", ")
 		}
-		fmt.Fprintf(&sb, "%s=[%s]", k, strings.Join(v, ", "))
+		fmt.Fprintf(&sb, "%s=[%s]", k, strings.Join(md[k], ", "))
 	}
 	fmt.Fprintf(&sb, "}")
 	return sb.String()
```

File: metadata/metadata.go
Comment (easwars) [inline metadata/metadata.go:94]: What do you think about sorting the keys before generating the string representation and ensuring that the output is deterministic. This will also make the unit test much simpler than what it is currently.

File: metadata/metadata.go
Comment (easwars) [inline metadata/metadata.go:99]: How about defining a `bool first` and using that as the conditional here, instead of this magic check which took me sometime to understand.

File: peer/peer_test.go
Comment (easwars) [reply peer/peer_test.go:53]: The use of `+` and `-` to convey the presence or absence of certain fields within the `Peer` isn't very readable. How about the following:
- withAddrAndValidAuth
- withAddrLocalAddrAndValidAuth
- withAddrAndEmptyAuth
- withAddrLocalAddrAndEmptyAuth
- nilPeer

The above names in the test output are more readable I feel.

**Comment by serathius** [comment]
Context why deterministic `String()` is important for etcd. 

Etcd uses grpc metadata to change properties of the bi-directional watch stream. Example of that is `WithRequireLeader()` which will ask server to break the stream if leader is no longer available in the cluster. As client can make multiple watch requests within a bi-directional stream, we need to create a separate stream for clients that provide different metadata in the request context. Etcd client get the identifiers of stream by formatting the context metadata, so it's crucial to have deterministic stringer to properly reuse the streams.

**Comment by easwars** [comment]
Phew.... Didn't realize that the PR was already approved and merged.

**Pull request merged**
